CI: build for Python 3.14, remove Python 3.9

rolandwalker opened the pull request:
Python 3.14 was released earlier this month.

This PR was carefully copy-pasted without understanding a few things.

It's not clear to me how the "Build wheels" PyO3/maturin-action would know how to find _all_ Python versions specified in setup-python/python-version, without something like `${{ matrix.python-version }}`.  Instead I would expect the build action to find the latest interpreter version.  But looking at https://pypi.org/project/theine-core/#files we see artifacts for many interpreters.

The Windows section had an extra `actions/setup-python` step for Python 3.13t so that was copied for 3.14t.

Worth noting but not addressed in this PR:
 * ~actions/setup-python is now on v6~ added to PR
 * ~Python 3.9 will be EOL in several days~ added to PR

rolandwalker committed:
CI: build for Python 3.14, remove Python 3.9

 * add Python 3.14
 * remove Python 3.9
 * update pyo3 to 0.27.1 in Cargo.toml
 * update lockfile
 * incidentally: upgrade actions/setup-python to v6

Yiling-J commented:
Thanks for the pull request! The CI failed because the `pyo3` version is outdated — we’ll need to update it to the latest version to make Python 3.14 build work. I think it’s fine to upgrade `setup-python` to v6 and drop Python 3.9. Feel free to include those changes in this PR if you’d like.

About your question: the `maturin-action` has a `--find-interpreter` parameter, which automatically detects all available Python versions on the host machine.

rolandwalker commented:
* updated setup-python to v6
* removed Python 3.9

PyO3/maturin-action is already on the latest version (v1 series), so I didn't immediately see how to update py03.  Maybe this PR is premature.

Yiling-J commented:
We actually need to upgrade PyO3 itself — it’s currently `0.24.2` in `Cargo.toml`. Update that line to

```
pyo3 = { version = "0.27.1", features = ["extension-module"] }
```

and then run `cargo update` to refresh `Cargo.lock`.

Can you try making that change and see if the CI passes?

rolandwalker commented:
> Update that line

Done, thanks!  All green.
